pkg/streamingpromql/operators/functions/sort.go

```diff
@@ -0,0 +1,168 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package functions
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"slices"
+
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type Sort struct {
+	Inner                    types.InstantVectorOperator
+	Descending               bool
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+
+	expressionPosition posrange.PositionRange
+
+	allData           []types.InstantVectorSeriesData // Series data, in the order returned by Inner
+	seriesOutputOrder []int                           // Series indices into allData, in the order they should be returned
+	seriesReturned    int                             // Number of series already returned by NextSeries
+}
+
+var _ types.InstantVectorOperator = &Sort{}
+
+func NewSort(
+	inner types.InstantVectorOperator,
+	descending bool,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	expressionPosition posrange.PositionRange,
+) *Sort {
+	return &Sort{
+		Inner:                    inner,
+		Descending:               descending,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+		expressionPosition:       expressionPosition,
+	}
+}
+
+func (s *Sort) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	innerSeries, err := s.Inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer types.PutSeriesMetadataSlice(innerSeries)
+
+	s.seriesOutputOrder, err = types.IntSlicePool.Get(len(innerSeries), s.MemoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+
+	s.allData = make([]types.InstantVectorSeriesData, len(innerSeries))
+
+	for idx := range innerSeries {
+		d, err := s.Inner.NextSeries(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		pointCount := len(d.Floats) + len(d.Histograms)
+
+		if pointCount > 1 {
+			return nil, fmt.Errorf("expected series %v to have at most one point, but it had %v", innerSeries[idx], pointCount)
+		}
+
+		if pointCount == 1 {
+			s.allData[idx] = d
+			s.seriesOutputOrder = append(s.seriesOutputOrder, idx)
+		} else {
+			types.PutInstantVectorSeriesData(d, s.MemoryConsumptionTracker)
+		}
+	}
+
+	if s.Descending {
+		slices.SortFunc(s.seriesOutputOrder, s.compareDescending)
+	} else {
+		slices.SortFunc(s.seriesOutputOrder, s.compareAscending)
+	}
+
+	// Why can't we just reorder innerSeries? This would be very difficult to do, as there's no
+	// guarantee that all the input series will be returned.
```

Comment:
I'm confused by this comment sorry?

Reply:
I think this is actually unnecessary, I think there's a better way I can do this that avoids the confusing comment as well 🙂 Will come back to this shortly.

Reply:
I've simplified the implementation in cc35cbe. 

Not only is this way simpler to understand, it performs just as well in our benchmarks.